src/unix/mmap.c

```diff
@@ -222,15 +222,49 @@ sysreturn mremap(void *old_address, u64 old_size, u64 new_size, int flags, void
     return sysreturn_from_pointer(vnew);
 }
 
+#if 0
+/* No-op callback for rangemap_find_gaps ... */
+static CLOSURE_0_1(vmap_gap, void, range);
+static void vmap_gap(range r) 
+{
+    rprintf("Found gap: [0x%lx, 0x%lx)\n", r.start, r.end);
+}
+#endif
+
 static sysreturn mincore(void *addr, u64 length, u8 *vec)
 {
-    if (validate_virtual(addr, length)) {
-        u32 vlen = pad(length, PAGESIZE) >> PAGELOG;
-        // presumably it wants the right valid bits set? - go doesn't seem to use it this way
-        for (int i = 0; i< vlen; i++) vec[i] = 1;
-        return 0;
+    u64 start, i;
+
+    start = u64_from_pointer(addr);
+    if (start & MASK(PAGELOG))
+        return -EINVAL;
+
+    length = pad(length, PAGESIZE);
+
+    /* XXX: we have no way to know whether there is unmapped memory in this range,
```

Comment:
It looks like perhaps this was part of an earlier, abandoned solution, but it's not clear to me why we even need to look at vmaps here. The page tables should tell us all we need to know, regardless of the type of mapping or meta.

The comment is correct for other purposes, however, in that we're not keeping a vmap for stuff like loaded ELF segments, the stack, heap, etc.. At some point - e.g. if we need to emulate /proc/self/maps - we should be able to easily add these. But I don't think it matters for mincore.

Reply:
mincore specifies that it wants -ENOMEM if anything in the specified range is not mapped. Conversely, if everything is mapped but some of the pages are not currently in RAM (e.g., not yet demand paged), fill the vec with 0. Unless I'm missing something, making the distinction between not mapped and not yet demand-paged required more info than we can get from the page tables -- e.g., a vmap

Reply:
Sorry, my mistake. Of course it wouldn't make sense for mincore to allow the inspection of a range of pages if the process didn't have it mapped. I think we can add those extra vmaps without much trouble.